onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,719 @@
+                                                    /* Level 2 */ {}};
+static std::unordered_set<std::string> allow_list; // Specified through configuration.
+
+static std::vector<std::string> compute_nodes_changed; // Names of the fp16_safe_ops changed
+
+// Check whether the given opcode is fp16 safe for the given level of optimization.
+static bool IsFP16Safe(const std::string& op_type, size_t level)
+{
+  bool fp16_safe = false;
+  for (size_t i=0; i < level && i < fp16_safe_ops.size() && !fp16_safe; ++i) {
+    fp16_safe = std::find(fp16_safe_ops[i].begin(), fp16_safe_ops[i].end(), op_type) != fp16_safe_ops[i].end();
+  }
+  return fp16_safe;
+}
+
+// Check whether the given opcode is fp16 allowed for the given level of optimization.
+static bool IsFP16Allow(const std::string& op_type, size_t level)
+{
+  bool fp16_allow = std::find(allow_list.begin(), allow_list.end(), op_type) != allow_list.end();
+  for (size_t i=0; i < level && i < fp16_allow_ops.size() && !fp16_allow; ++i) {
+    fp16_allow = std::find(fp16_allow_ops[i].begin(), fp16_allow_ops[i].end(), op_type) != fp16_allow_ops[i].end();
+  }
+  return fp16_allow;
+}
+
+static bool IsCastTo(const Node& node, TensorProto_DataType data_type)
+{
+  if (node.OpType() == "Cast") {
+    const NodeAttributes& attributes = node.GetAttributes();
+    ORT_ENFORCE(attributes.find("to") != attributes.end());
+    return attributes.at("to").i() == static_cast<int64_t> (data_type);
+  }
+  return false;
+}
+
+static bool IsType(const NodeArg& node_arg, TensorProto_DataType data_type)
+{
+  return node_arg.TypeAsProto()->tensor_type().elem_type() == data_type;
+}
+
+// InsertCastNodes
+// Insert a new Cast node after each NodeArg in the require_cast vector. The cast node is FLOAT16 if is_fp16 is True
+// and FLOAT otherwise. This funtion fixes the graph edges in addition to inserting the cast nodes.
+static Status InsertCastNodes(Graph& graph,
+                              const std::unordered_set<NodeArg*>& require_cast,
+                              bool is_fp16,
+                              std::deque<onnxruntime::NodeIndex>& removed_nodes)
+{
+  //Create requirred new Cast nodes.
+  for (NodeArg* node_arg : require_cast) {
+    if (!node_arg->Exists()) {
+      continue;
+    }
+    // data_type is the data type of the Cast output.
+    TensorProto_DataType data_type = is_fp16 ? TensorProto_DataType_FLOAT16 : TensorProto_DataType_FLOAT;
+    TypeProto type_proto;
+    bool is_node_arg_cast_output = IsType(*node_arg, data_type);
+    TensorProto_DataType new_node_arg_data_type = data_type;;
+    if (is_node_arg_cast_output) {
+      new_node_arg_data_type = (data_type == TensorProto_DataType_FLOAT) ? TensorProto_DataType_FLOAT16 : TensorProto_DataType_FLOAT;
+    }
+    type_proto.mutable_tensor_type()->set_elem_type(new_node_arg_data_type);
+    NodeArg& new_node_arg = graph.GetOrCreateNodeArg(graph.GenerateNodeArgName(node_arg->Name()), &type_proto);
```

Comment:
you can just leave type_proto as nullptr and the type inference logic will figure it out

Reply:
I have many functions operating on this graph. I am only resolving at the top of the do-while loop in ApplyImpl. I want to keep it consistent

---
In reply to: [610857809](https://github.com/microsoft/onnxruntime/pull/7191#discussion_r610857809) [](ancestors = 610857809)